File: scripts/cleanup-inactive-users.py
#!/usr/bin/env python3
"""
清理长期不活跃用户 + 过期验证码
策略：
- 2 年未登录的用户标记为 suspended + 清理 session
- 30 天以上的验证码自动清理
每月 1 号 04:00 执行
"""
import sqlite3, os, json, time
from pathlib import Path

DB_PATH = Path(__file__).parent.parent / "apps" / "auth-service" / "data" / "auth.db"
LOG_FILE = Path(__file__).parent.parent / "backup" / "db" / "cleanup.log"
NOW = time.strftime("%Y-%m-%d %H:%M:%S")

if not DB_PATH.exists():
    print(f"[{NOW}] auth.db not found: {DB_PATH}")
    exit(1)

conn = sqlite3.connect(str(DB_PATH))
c = conn.cursor()

# 1. 标记 2 年不活跃用户
two_years_ago = time.strftime("%Y-%m-%d %H:%M:%S", time.localtime(time.time() - 730 * 86400))
stale = c.execute(
    "SELECT id, username, nickname FROM users WHERE updated_at < ? AND suspended = 0",
    (two_years_ago,)
).fetchall()

if stale:
    ids = [row[0] for row in stale]
    placeholders = ",".join("?" * len(ids))
    c.execute(f"UPDATE users SET suspended = 1, updated_at = datetime('now') WHERE id IN ({placeholders})", ids)
    c.execute(f"DELETE FROM sessions WHERE user_id IN ({placeholders})", ids)
    print(f"[{NOW}] suspended {len(stale)} stale users: {[r[1] or r[0] for r in stale[:5]]}...")
else:
    print(f"[{NOW}] no stale users found")

# 2. 清理过期验证码
del_codes = c.execute("DELETE FROM verification_codes WHERE created_at < datetime('now', '-30 days')").rowcount
print(f"[{NOW}] cleaned {del_codes} expired codes")

conn.commit()
conn.close()

# 3. 写日志
LOG_FILE.parent.mkdir(parents=True, exist_ok=True)
old_logs = ""
if LOG_FILE.exists():
    old_logs = "".join(LOG_FILE.read_text(errors="replace").splitlines(keepends=True)[:99])
LOG_FILE.write_text(f"[{NOW}] stale={len(stale)} codes_clean={del_codes}\n" + old_logs)

print("done")

File: scripts/cleanup-inactive-users.mjs
#!/usr/bin/env node
/**
 * 清理长期不活跃用户
 * 策略：2 年未登录 → 标记 suspended + 删除 session
 *       过期验证码 → 30 天以上清理
 * 每月 1 号 04:00 执行
 */

import { readFileSync, writeFileSync, existsSync } from 'fs'
import { resolve, dirname } from 'path'
import { fileURLToPath } from 'url'

const __dirname = dirname(fileURLToPath(import.meta.url))

async function main() {
  const dbPath = resolve(__dirname, '../apps/auth-service/data/auth.db')
  if (!existsSync(dbPath)) {
    console.error('auth.db not found:', dbPath)
    process.exit(1)
  }

  const betterSqlite3 = await import(resolve(__dirname, '../apps/auth-service/node_modules/better-sqlite3/lib/index.js'))
  const conn = betterSqlite3.default(dbPath)
  const logFile = resolve(__dirname, '../backup/db/cleanup.log')
  const now = new Date().toISOString()
  const twoYearsAgo = new Date(Date.now() - 730 * 24 * 60 * 60 * 1000).toISOString()

  // 1. 标记 2 年不活跃用户 + 清理 session
  const staleUsers = conn.prepare(
    `SELECT id, username, nickname FROM users WHERE updated_at < ? AND suspended = 0`
  ).all(twoYearsAgo)

  console.log(`[${now}] stale users: ${staleUsers.length}`)
  if (staleUsers.length > 0) {
    const ids = staleUsers.map(u => u.id)
    const placeholders = ids.map(() => '?').join(',')

    conn.prepare(`UPDATE users SET suspended = 1, updated_at = datetime('now') WHERE id IN (${placeholders})`).run(...ids)
    conn.prepare(`DELETE FROM sessions WHERE user_id IN (${placeholders})`).run(...ids)

    staleUsers.forEach(u => console.log(`  suspended: ${u.username || u.id}`))
  }

  // 2. 清理过期验证码
  const delCodes = conn.prepare(`DELETE FROM verification_codes WHERE created_at < datetime('now', '-30 days')`).run()
  console.log(`  codes cleaned: ${delCodes.changes}`)

  // 3. 写日志
  let logContent = `[${now}] stale=${staleUsers.length} codes_clean=${delCodes.changes}\n`
  try {
    if (existsSync(logFile)) {
      logContent += readFileSync(logFile, 'utf-8').split('\n').slice(0, 99).join('\n')
    }
    writeFileSync(logFile, logContent)
  } catch {}

  conn.close()
  console.log('done')
}

main().catch(e => { console.error('error:', e.message); process.exit(1) })